File: LeetCode/Diameter_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int height(TreeNode *root , int *dia){\
        int left ,right;
        if(!root) return 0;
        else{
            left = height(root->left,dia);
            right = height(root->right,dia);
            int a = left+ right;
            if((*dia)  < a)
                *dia = a;                
        }
        return max(left,right)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int dia = 0;
        height(root,&dia);
        return dia;
    }
};

File: LeetCode/diameter_of_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private static  int  diameter = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        Solution.diameter = 0;
        calculateDiameter(root);
        return Solution.diameter;
    }
    
    public int calculateDiameter(TreeNode root){
        if(root == null) return 0;
        else{
            int left = calculateDiameter(root.left);
            int right = calculateDiameter(root.right);
            
            if(Solution.diameter < left+right){
                Solution.diameter = left+right;
                //System.out.println(" Diameter is updated at " + root.val + " to " + Solution.diameter);
            }
            return Integer.max(left, right) + 1;
        }
    }
}